addons/udes_stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -1,5 +1,581 @@
+            self.write({"u_last_reserved_pallet_name": pallet_name})
+
+    def _get_task_grouping_criteria(self):
+        """
+        Return a function for sorting by picking, package(maybe), product, and
+        location. The package is not included if the picking type allows for
+        the swapping of packages (`u_allow_swapping_packages`) and picks by
+        product (`u_user_scans`)
+        """
+        batch_pt = self.mapped("picking_ids.picking_type_id")
+        batch_pt.ensure_one()
+
+        parts = [lambda ml: (ml.picking_id.id,)]
+
+        if not (batch_pt.u_allow_swapping_packages and batch_pt.u_user_scans == "product"):
+            parts.append(lambda ml: (ml.package_id.id,))
+
+        parts.append(lambda ml: (ml.location_id.id, ml.product_id.id))
+
+        return lambda ml: tuple(chain(*[part(ml) for part in parts]))
+
+    def get_available_move_lines(self):
+        """ Get all the move lines from available pickings
+        """
+        self.ensure_one()
+        available_pickings = self.picking_ids.filtered(lambda p: p.state == "assigned")
+
+        return available_pickings.move_line_ids
+
+    def get_next_tasks(
+        self,
+        skipped_product_ids=None,
+        skipped_move_line_ids=None,
+        task_grouping_criteria=None,
+        limit=1,
+    ):
+        """
+        Get the next not completed tasks of the batch to be done.
+        Expect a singleton.
+
+        Note that the criteria for sorting and grouping move lines
+        (for the sake of defining tasks) are given by the
+        _get_task_grouping_criteria method so it can be specialized
+        for different customers. Also note that the
+        task_grouping_criteria argument is added to the signature to
+        enable dependency injection for the sake of testing.
+
+        Confirmations is a list of dictionaries of the form:
+            {'query': 'XXX', 'result': 'XXX'}
+        After the user has picked the move lines, should be requested by the
+        'query' to scan/type a value that should match with 'result'.
+        They are enabled by picking type and should be filled at
+        _prepare_task_info(), by default it is not required to confirm anything.
+        """
+        MoveLine = self.env["stock.move.line"]
+
+        self.ensure_one()
+
+        all_available_mls = self.get_available_move_lines()
+        skipped_mls = MoveLine.browse()
+
+        # Filter out skipped move lines
+        if skipped_product_ids:
+            skipped_mls = all_available_mls.filtered(
+                lambda ml: ml.product_id.id in skipped_product_ids
+            )
+        elif skipped_move_line_ids:
+            skipped_mls = all_available_mls.filtered(lambda ml: ml.id in skipped_move_line_ids)
+        available_mls = all_available_mls - skipped_mls
+
+        num_tasks_picked = len(available_mls.filtered(lambda ml: ml.qty_done == ml.product_qty))
+
+        todo_mls = available_mls.get_lines_todo().sort_by_location_product()
```

Comment:
Can we call a method that calls sort_by_location_product() instead of directly calling it? this will make the code easier to extend. Same for line 319.